fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/RewriteMultiDistinctOptimizedRule.java

```diff
@@ -0,0 +1,179 @@
+    }
+
+    @Override
+    public boolean check(OptExpression input, OptimizerContext context) {
+        LogicalAggregationOperator agg = (LogicalAggregationOperator) input.getOp();
+
+        List<CallOperator> distinctAggOperatorList = agg.getAggregations().values().stream()
+                        .filter(CallOperator::isDistinct).collect(Collectors.toList());
+
+        boolean hasMultiColumns = false;
+        for (CallOperator callOperator : distinctAggOperatorList) {
+            if (callOperator.getChildren().size() > 1) {
+                hasMultiColumns = true;
+                break;
+            }
+        }
+
+        return (distinctAggOperatorList.size() > 1 && !hasMultiColumns) || agg.getAggregations().values().stream()
+                .anyMatch(call -> call.isDistinct() && call.getFnName().equals(FunctionSet.AVG));
+    }
+
+    @Override
+    public List<OptExpression> transform(OptExpression input, OptimizerContext context) {
+        ColumnRefFactory columnRefFactory = context.getColumnRefFactory();
+        // define cteId
+        Integer cteId = columnRefFactory.getNextRelationId();
+        // build logic cte produce operator
+        List<OptExpression> aggChildOptExpressionList = input.getInputs();
+        LogicalCTEProduceOperator logicalCTEProduceOperator = new LogicalCTEProduceOperator(cteId);
+        // build cte anchor operator
+        LogicalCTEAnchorOperator anchorOperator = new LogicalCTEAnchorOperator(cteId);
+        List<OptExpression> anchorOperatorInputsList = new ArrayList<>();
+        anchorOperatorInputsList.add(OptExpression.create(logicalCTEProduceOperator, aggChildOptExpressionList));
+        // get all aggregation operator
+        LogicalAggregationOperator originalAllAggregationOperator = (LogicalAggregationOperator) input.getOp();
+        Map<ColumnRefOperator, CallOperator> countColumnMap = originalAllAggregationOperator.getAggregations();
+        Map<String, ColumnRefOperator> columnMap = new HashMap<>();
+        for (Map.Entry<ColumnRefOperator, CallOperator> entry : countColumnMap.entrySet()) {
+            ColumnRefOperator crf = (ColumnRefOperator) entry.getValue().getChild(0);
+            String columnName = crf.getName();
+            columnMap.put(columnName, crf);
+        }
+
+        Map<ColumnRefOperator, CallOperator> copyCountColumnMap = new HashMap<>();
+        copyCountColumnMap.putAll(countColumnMap);
+        OptExpression inputs = constructJoinOperatorInputs(null, copyCountColumnMap, columnMap,
+                columnRefFactory,
+                originalAllAggregationOperator,
+                cteId);
+
+
+        anchorOperatorInputsList.add(inputs);
+
+        context.getSessionVariable().setCboCteReuse(true);
+        context.getCteContext().setEnableCTE(true);
+        // add forced identifier
+        context.getCteContext().setForcedCTE(true);
+        context.getSessionVariable().setEnablePipelineEngine(true);
```

Comment:
check there flag in `check` method, don't execute the rule if there flag is false